### ISSUE TITLE
terraform remote state not being read properly

### ISSUE THREAD

**Issue by ura718** [hashicorp/terraform#20537]
I am seeing a weird scenario where I am storing a vpc.tfstate file in s3 that has all the necessary information but when I am referencing it its not being picked up.

```
$ terraform version
Terraform v0.11.11
+ provider.aws v2.0.0
```

Location of vpc.tfstate stored in s3
```
$ terraform state show terraform_remote_state.vpc
id            = 2019-02-28 21:26:54.469504658 +0000 UTC
backend       = s3
config.%      = 3
config.bucket = terraform
config.key    = env/prod/us-east-1/common/vpc/vpc.tfstate
config.region = us-east-1
environment   = default
workspace     = default
```

Copying vpc.tfstate to local disk to view whats inside. It has a lot of info inside so I am grepping for stuff that I get errors on. Looks like it has the necessary info that I need. The `vpc_id` and `subnet_public_a`.
```
$ aws s3 cp s3://terraform/env/prod/us-east-1/common/vpc/vpc.tfstate .
download: s3://terraform/env/prod/us-east-1/common/vpc/vpc.tfstate to ./vpc.tfstate

$ grep vpc_id vpc.tfstate
                "vpc_id": {
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"
                            "vpc_id": "vpc-97bb1ef2"

$ grep -A 5 subnet_public_a vpc.tfstate
                "subnet_public_a": {
                    "sensitive": false,
                    "type": "string",
                    "value": "subnet-261a1d0e"
                },
```

Snippet of terraform code that I run that supposed to read `terraform_remote_state` file and get info from there. But its not picking up the information from `vpc.tfstate` file even though I know the `vpc.tfstate` file is good.

```
### ------------------------------------------------
### AWS Region VPC uses
### ------------------------------------------------
provider "aws" {
  region  = "us-east-1"
  profile = "virginia-prod"
}




### ------------------------------------------------
### Read Only reference to vpc state
### when referencing data from vpc.tfstate make sure
### vpc module has outputs so you can reference them
### from this code base using data.terraform_remote_state.vpc...
### ------------------------------------------------
data "terraform_remote_state" "vpc" {
    backend = "s3"
    config {
      bucket = "terraform"
      key    = "env/prod/us-east-1/common/vpc/vpc.tfstate"
      region = "us-east-1"
    }
}




### ------------------------------------------------
### Store bastion in S3 tfstate
### ------------------------------------------------
terraform {
  backend "s3" {
    bucket = "terraform"
    key    = "env/prod/us-east-1/bastion/bastion.tfstate"
    region = "us-east-1"
  }
}



### ------------------------------------------------
### Retrieve ami id from aws
### ------------------------------------------------
data "aws_ami" "linux" {

  most_recent = true

  filter {
    name   = "name"
    values = ["amzn2-ami-hvm-*-x86_64-gp2", ]
  }


  filter {
    name   = "owner-alias"
    values = [ "amazon" ]
  }

  owners = ["amazon"]
}




### ------------------------------------------------
### Setup bastion host
### ------------------------------------------------
module "bastion" {
  source = "../../../modules/bastion"

  vpc_id         = "${data.terraform_remote_state.vpc.vpc_id}"

  ami            = "${data.aws_ami.linux.id}"
  key_name       = "aws-ec2user-global-key.pub"
  instance_type  = "t2.micro"
  count          = "1"
  subnet_id      = "${data.terraform_remote_state.vpc.subnet_public_a}"

  # Tags
  name           = "Bastion"
  buildby        = "Terraform"
```

Here is the module code snippet information
```
$ cat ../../../modules/bastion/bastion.tf
### ------------------------------------------------
### Setup bastion host instance
### ------------------------------------------------
resource "aws_instance" "bastion" {
  ami                         = "${var.ami}"
  key_name                    = "${aws_key_pair.bastion_key.key_name}"
  instance_type               = "${var.instance_type}"
  count                       = "${var.count}"
  subnet_id                   = "${var.subnet_id}"
  #security_groups            = ["${aws_security_group.bastion-sg.id}"]   // will destroy/create new ec2 each time terraform apply
  vpc_security_group_ids      = ["${aws_security_group.bastion-sg.id}"]   // will update in place without destroy/create
  associate_public_ip_address = true

  tags {
    Name  = "${var.name}"
    Build = "${var.buildby}"
  }


  lifecycle {
    create_before_destroy = true
  }

...
...
...

### ------------------------------------------------
### Setup bastion security group
### ------------------------------------------------
resource "aws_security_group" "bastion-sg" {
  #name_prefix   = "${var.name}-sg"                  // Resolves duplicate name when creating and deleting resource through lifecycle
  name           = "${var.name}-sg"                  // Resolves duplicate name when creating and deleting resource through lifecycle
  description    = "Open port 22 to bobst network"
  vpc_id         = "${var.vpc_id}"


  ingress {
    protocol     = "${var.protocol}"
    from_port    = "${var.from_port}"
    to_port      = "${var.to_port}"
    cidr_blocks  = ["${var.cidr_blocks}"]
  }

  egress {
    protocol     = -1
    from_port    = 0
    to_port      = 0
    cidr_blocks  = ["0.0.0.0/0"]
  }

  tags {
    Name         = "${var.name}-sg"
    Build        = "${var.buildby}"
  }


  lifecycle {
    create_before_destroy = true
  }
}
```

Here is a snippet of the `vars.tf` file in modules directory

```
$ cat ../../../modules/bastion/vars.tf
 ### --------------------
### bastion.tf
### --------------------
variable "vpc_id"            { description = "Vpc Id" }
variable "ami"               { description = "AMI ID" }
variable "key_name"          { description = "Name of public ssh key" }
variable "instance_type"     { description = "Instance Type e.g: t2.micro" }
variable "count"             { description = "Number of EC2 instance to launch" }
variable "subnet_id"         { description = "Public Subnet Ids" }
```

Finally running terraform plan produces this error.
```
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.terraform_remote_state.vpc: Refreshing state...
data.aws_ami.linux: Refreshing state...

------------------------------------------------------------------------

Error: Error running plan: 2 error(s) occurred:

* module.bastion.var.vpc_id: Resource 'data.terraform_remote_state.vpc' does not have attribute 'vpc_id' for variable 'data.terraform_remote_state.vpc.vpc_id'
* module.bastion.var.subnet_id: Resource 'data.terraform_remote_state.vpc' does not have attribute 'subnet_public_a' for variable 'data.terraform_remote_state.vpc.subnet_public_a'
```

This is because in the previews showcase the `$ terraform state show terraform_remote_state.vpc` showed empty vpc.tfstate file even though it has the necessary information. So my question is what is it that I am doing wrong such that its not picking up the vpc.tfstate file from s3 bucket?

Thank you.

**Comment by apparentlymart** [comment]
Hi @ura718,

In the configuration that is generating the `vpc` and `bastion` remote states here, do you have `output` blocks declaring which values you want to export? I didn't see any in your examples here, but I'm not sure if that's just because you excluded them when summarizing.

**Comment by ura718** [comment]
I do have outputs, here is my output.tf file in modules directory. 

```
output "vpc_id" {
  value = "${aws_vpc.vpc.id}"
}


output "subnet_public_ids"{
  value = [
    "${aws_subnet.public_a.id}",
    "${aws_subnet.public_b.id}"
  ]
}


output "subnet_private_ids"{
  value = [
    "${aws_subnet.private_a.id}",
    "${aws_subnet.private_b.id}"
  ]
}


output "subnet_public_a"{
  value = "${aws_subnet.public_a.id}"
}

output "subnet_public_b"{
  value = "${aws_subnet.public_b.id}"
}
```

However, I believe the problem is such that its not reading or picking up  the `terraform_remote_state.vpc` block from S3, and so it does not have the necessary information to output. Strangely enough this entire setup works in my development since the code is very similar with just some tweaks to point to different regions. But in production its not working. I even tried to remove the `.terraform` directory and redo `terraform init` so it gets the plugins and loads the module info but its not getting that vpc.tfstate from s3 properly. Here is an example of development vs production:

Development:
```
$ terraform state show terraform_remote_state.vpc
id                   = 2019-02-28 20:26:32.857922221 +0000 UTC
backend              = s3
config.%             = 3
config.bucket        = dev-terraform
config.key           = env/dev/us-east-1/common/vpc/vpc.tfstate
config.region        = us-east-2
environment          = default
subnet_private_ids.# = 2
subnet_private_ids.0 = subnet-3724bb40
subnet_private_ids.1 = subnet-918a35c8
subnet_public_a      = subnet-4824bb3f
subnet_public_b      = subnet-908a35c9
subnet_public_ids.#  = 2
subnet_public_ids.0  = subnet-4824bb3f
subnet_public_ids.1  = subnet-908a35c9
vpc_id               = vpc-883771ed
workspace            = default
``` 

Production
```
$ terraform state show terraform_remote_state.vpc
id            = 2019-02-28 21:26:54.469504658 +0000 UTC
backend       = s3
config.%      = 3
config.bucket = terraform
config.key    = env/prod/us-east-1/common/vpc/vpc.tfstate
config.region = us-east-1
environment   = default
workspace     = default
```

**Comment by apparentlymart** [comment]
Thanks for confirming, @ura718!

To better understand what's going on here, it would help to have a trace log. Could you make a Terraform configuration containing _only_ the `data.terraform_remote_state.vpc` block and run `terraform plan` with the environment variable `TF_LOG=trace` set, and then put the resulting verbose output in a [gist](https://gist.github.com/)?

The log output should help us to see exactly what `terraform_remote_state` is requesting and what result it got. Of course, feel free to delete from the log the full resource details and any S3 API request headers you don't want to share; I'm mainly interested to see the S3 request URL/body and response status/body to see if there's anything unusual.

**Comment by ura718** [comment]
Certainly, here is the link to trace log

`https://gist.github.com/ura718/0dd2f582331ddbf7881e4671025afe70`

Also here is the code snippet of the `bastion.tf` stripped down to only access s3.

```
### ------------------------------------------------
### AWS Region VPC uses
### ------------------------------------------------
provider "aws" {
  region  = "us-east-1"
  profile = "virginia-prod"
}




### ------------------------------------------------
### Read Only reference to vpc state
### when referencing data from vpc.tfstate make sure
### vpc module has outputs so you can reference them
### from this code base using data.terraform_remote_state.vpc...
### ------------------------------------------------
data "terraform_remote_state" "vpc" {
    backend = "s3"
    config {
      bucket = "terraform"
      key    = "env/prod/us-east-1/common/vpc/vpc.tfstate"
      region = "us-east-1"
    }
}




### ------------------------------------------------
### Store bastion in S3 tfstate
### ------------------------------------------------
terraform {
  backend "s3" {
    bucket = "terraform"
    key    = "env/prod/us-east-1/bastion/bastion.tfstate"
    region = "us-east-1"
  }
}
```

Also here is the output of terraform state list and state show, and I see that the configs now have appended numbers to them (not sure why), which is different from original post.
```
$ terraform state list
terraform_remote_state.vpc


$ terraform state show terraform_remote_state.vpc
id                       = 2019-03-01 19:52:10.527684593 +0000 UTC
backend                  = s3
config.#                 = 1
config.3979633902.bucket = terraform
config.3979633902.key    = env/prod/us-east-1/common/vpc/vpc.tfstate
config.3979633902.region = us-east-1
environment              = default
workspace                = default
```

**Comment by ura718** [comment]
Ok, so I found the problem and what was causing the issue with production vpc.tfstate. This was a mistake on my part and when you initially asked me if I had the outputs for vpc setup I mentioned that I did. However, I completely missed the fact that they were commented out as one large block in the file with `/*` and `*/` causing variables not to populate into the `terraform_remote_state.vpc`. Once I removed the comments everything started to work as normal. Sorry for the confusion and going down a rabbit hole where nothing made sense. This felt similar to forgetting to put a semicolon in a c++ program and wondering why the code did not work. Thank you for the help, the world is back to normal again.
